Source Upload Form 컴포넌트 및 기능 구현

Jeongeun-Choi opened the pull request:
## 📕 제목

Source Upload Form 컴포넌트 및 기능 구현
관련이슈 #4, #21 

## 📗 작업 내용

> 구현 내용 및 작업 했던 내역

- [x] 로컬에서 오디오 파일을 불러온다.
- [x] 불러오기 영역의 드래그 앤 드랍으로 파일을 불러올 수 있다.
- [x] 불러오기 영역을 클릭해서 로컬파일을 직접 선택하여 불러올 수 있다.
- [x] 소스불러오기는 클릭, 드래그 앤 드랍 모두 가능하다

기본 화면
![image](https://user-images.githubusercontent.com/55783203/100088912-39853200-2e94-11eb-99ba-e6b37d394629.png)

소스 불러왔을 때
![image](https://user-images.githubusercontent.com/55783203/100089026-62a5c280-2e94-11eb-9b19-d9e29b4c831d.png)

## 📘 PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

- 일단은 하나의 오디오 파일만 불러오도록 구현했습니다. 나중에 여러개를 불러오고싶다면 확장해도 될 것 같습니다. 
- 드래그 앤 드롭시 드래그 중일땐 배경색이 다르도록 구현했으나 적용이 안되네요. 추후에 고치겠습니다.
- 비디오 파일도 오디오 추출이 가능하다는 것을 확인했습니다! 하지만 지금 코드에선 audio 파일만 올릴 수 있도록 설정을 해둬서 추후에 비디오도 올릴 수 있도록 하겠습니다.

Jeongeun-Choi committed:
feat: [FE] Modal Close Event 구현

- 본문 내용
Modal 배경화면을 누르거나 취소 버튼을 눌렸을때 Modal 창이 닫히는 함수를
구현했습니다.

Jeongeun-Choi committed:
style: [FE] Modal Style 추가

- 본문 내용
 - Modal conent가 정중앙에 오도록 구현했습니다.
 - 자주 쓰일 것 같은 색상을 따로 변수로 두었습니다.

Jeongeun-Choi committed:
feat: [FE] source-upload-form 컴포넌트 생성

- 본문 내용
  - 드래그 앤 드롭으로 파일을 가져올 수 있습니다.
  - Click or Drag and Drop 텍스트를 눌러도 파일을 가져올 수 있습니다.

Jeongeun-Choi committed:
style: [FE] source-upload-form style

- 본문 내용
source-uplaod-form과 관련된 style을 작성했습니다.

Jeongeun-Choi committed:
feat: [FE] Analyzer 컴포넌트 작성

- 본문 내용
ArrayBuffer를 받아서 AudioBuffer로 변환시켜주는 메소드가 작성되어 있다.

Songwonseok commented on frontend/src/components/modal/index.ts:
```diff
@@ -0,0 +1,17 @@
+
+  closeModal(e): void {
+    const { target } = e;
+    const modalElement: HTMLElement | null = document.querySelector('.modal');
```
오 이런식으로 type을 먹일 수 있겠네요. 배워갑니다!

wooojini commented on frontend/src/components/audio-analyzer/index.ts:
```diff
@@ -0,0 +1,17 @@
+  setAudio(audioFile) {
+    this.audioContext.decodeAudioData(audioFile).then((buffer: AudioBuffer) => {
+      this.audioBuffer = buffer;
+      console.log(this.audioBuffer);
```
앗! console은 지워주세요 :-)

Jeongeun-Choi replied:
넵 지웠습니다!

wooojini commented on frontend/src/components/modal/style.scss:
```diff
@@ -0,0 +1,28 @@
+$greenColor: #03c75a;
+
+.modal {
+  display: flex;
```
CSS 정렬을 수행하는 것은 어떨까요??
참고 :  https://webdesign.tutsplus.com/articles/outside-in-ordering-css-properties-by-importance--cms-21685

Jeongeun-Choi replied:
오 css도 정렬이 필요하군요... 막 썼는데 ㅎㅎ;; 감사합니다!

wooojini commented on frontend/src/components/source-upload/index.ts:
```diff
@@ -0,0 +1,83 @@
+    constructor() {
+      super();
+      this.filename = '';
+      this.sourceUploadElement = document.querySelector('.source-upload-content');
```
생성자에서 DOM 접근이 가능하셨나요??! 생성자가 호출되는 타이밍에서는 아직 render 함수도 안불리고 해당 DOM 접근이 안될 것 같은데 궁금합니다!! :-)

Jeongeun-Choi replied:
해당 클래스 안의 함수들에서 같은 DOM을 접근하는 코드가 중복이 되어서 따로 빼놨었는데 호출되는 타이밍에서  null로 값이 들어가서 실행이 안 될 줄 알았는데 되긴하더라구요😅 
갑자기 생각난건데 저걸 connectedCallback 함수에서 정의하면 접근이 잘 될지 궁금하네요ㅎㅎ
한 번 도전해봐야겠습니다...

Jeongeun-Choi committed:
style: [FE] Style 속성 제거

- 본문 내용
Modal 관련 css를 제거했습니다.

wooojini commented on frontend/src/components/source-upload/style.scss:
```diff
@@ -0,0 +1,47 @@
+
+.source-upload-buttons {
+  @extend %modal-buttons;
+
```
SCSS를 모듈화해서 잘 사용해주셨네요 :-)

wooojini reviewed (approved):
고생하셨습니다!! 전체적으로 잘해주셨네요!

Jeongeun-Choi committed:
fix: [FE] console.log 삭제

- 본문 내용
console.log 삭제 했습니다.

Jeongeun-Choi committed:
style: [FE] css 정렬

- 본문 내용
css 정렬을 했습니다!

pieisland commented on frontend/src/components/modal/style.scss:
```diff
@@ -0,0 +1,28 @@
+  width: 50%;
+  height: 60%;
+  text-align: center;
+  border: 1px solid #000000;
+  background-color: $baseColor;
+  border: 2px solid $borderColor;
```
border가 두 번 쓰여있어서 하나를 지우시면 될 것 같습니다.

Jeongeun-Choi replied:
앗 알려주셔서 감사합니다!

pieisland commented:
파일 업로드까지 빠르게 해주셨네요! web audio api에 대해 이해를 잘 하고 계신 것 같아요 ㅎ.ㅎ 고생많으셨어요!!

Jeongeun-Choi committed:
fix: [FE] 코드 분리

- 본문 내용
  - initDOM 함수를 만들어 DOM을 탐색하는 함수를 만들었습니다.
  - initEvent 함수를 만들어 Event를 등록하는 함수를 만들었습니다.

Jeongeun-Choi committed:
refactor: 사용하지 않는 코드 삭제

- 본문 내용
사용하지 않는 코드를 삭제했습니다.

wooojini committed:
Merge branch 'develop' into feature/source-upload-form
